### PR TITLE
[builder] Support for --skip-new-go-module

### PR DESCRIPTION
**Pull request by kristinapathak**
A continuation of https://github.com/open-telemetry/opentelemetry-collector/pull/9253

Description: Adds a --skip-new-go-module flag to the OTC builder. This enables users working in an existing go module environment (say, a "monorepo") to update the module they have, vs forcing the use of a new module.

With the new support inside an existing Go module, a collector main package can be generated using a go:generate directive. For example, in the directory where I want my collector built, the file generate.go has this line:

//go:generate builder --skip-new-go-module --skip-compilation --strict-versioning --config=./build-config.yaml
In the same directory, the build-config.yaml describes the collector to build. The builder generates the other files in the same directory. At this point, normal Go workflows can be used to update indirect dependencies. The optional --strict-versioning checks that the build configuration matches that resulting from the enclosing Go module.

Link to tracking Issue: https://github.com/open-telemetry/opentelemetry-collector/issues/9252

Testing: Unit tests added.

Documentation: Yes.

### CODE REVIEW

**Comment by kristinapathak** [comment]
Discussed at the collector sig meeting today - I will look into splitting up the strict versioning and skip new go module flag additions into their own PRs.

**Comment by jpkrohling** [comment]
>  I will look into splitting up the strict versioning and skip new go module flag additions into their own PRs.

Please ping me directly once you have them ready, I'll review the PRs.

File: cmd/builder/README.md [hunk 1]
```diff
@@ -145,3 +145,44 @@ then commit the code in a git repo. A CI can sync the code and execute
...
+1. The `dist::otelcol_version` field in the build configuration must
+   match the core library version calculated by the Go toolchain,
+   considering all components.  A mismatch could happen, for example,
+   when one of the components depends on a newer release of the core
+   collector library.
+2. For each component in the build configuration, the version included
+   in the `gomod` module specifier must match the one calculated by
+   the Go toolchain, considering all components.  A mismatch could
+   happen, for example, when the enclosing Go module uses a newer
+   release of the core collector library.
```
Comment (mx-psi) [inline cmd/builder/README.md:185]: I think this should be the default, and would have prevented issues such as #8692. Let's do this on a separate PR, I am happy to review that one :)

**Comment by kristinapathak** [comment]
I'll be rebasing this and adding tests tonight.

**Comment by kristinapathak** [comment]
Declaring PR bankruptcy and trying again [here](https://github.com/open-telemetry/opentelemetry-collector/pull/10098).
